compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -1309,7 +1396,10 @@ object Types extends TypeUtils {
     final def widen(using Context): Type = this match
       case _: TypeRef | _: MethodOrPoly => this // fast path for most frequent cases
       case tp: TermRef => // fast path for next most frequent case
-        if tp.isOverloaded then tp else tp.underlying.widen
+        // Don't call `isOverloaded` and `underlying` on `tp` directly,
+        // to avoid computing provisional state twice.
+        val denot = tp.denot
+        if denot.isOverloaded then tp else denot.info.widen
```

Comment:
I think we can rely on `underlying.exist` for `TermRef`, but not for other `SingletonType`s.

Reply:
Yes, I was thinking about the `widenSingleton`, we can add a case for `TermRef`.

Reply:
Oh, it seems we cannot do this. There is case where tp is not overloaded but underlying does not exist.

Reply:
I see, you're right.  But it feels to me like the isOverloaded guard is there exactly to avoid TermRef's NoType result from being the widen{,Singleton} result.  And thus returning `tp`, even when a non-overloaded TermRef has no underlying type, is proper.

But that's all speculative - let's stick to your fix.